eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -1,132 +1,430 @@
+                  , Set.map unGovPurposeId hardForkRC
+                  , Set.map unGovPurposeId committeeRC
+                  , Set.map unGovPurposeId constitutionRC
+                  ]
+            sumOfSizes =
+              sum
+                [ Set.size pparamUpdateRC
+                , Set.size hardForkRC
+                , Set.size committeeRC
+                , Set.size constitutionRC
+                ]
+         in sizeOfUnions == sumOfSizes
+      )
+    $ assert
+      -- The union of the four hierarchies should be equal to the OMap
+      (Map.keysSet (OMap.toMap omap) == unionGraph)
+    $ assert
+      -- Root-children should be subsets of the hierarchy
+      ( and
+          [ pparamUpdateRC `Set.isSubsetOf` Map.keysSet pparamUpdateH
+          , hardForkRC `Set.isSubsetOf` Map.keysSet hardForkH
+          , committeeRC `Set.isSubsetOf` Map.keysSet committeeH
+          , constitutionRC `Set.isSubsetOf` Map.keysSet constitutionH
+          ]
+      )
+    $ assert
+      -- Children of nodes should be subsets of the hierarchy
+      ( and
+          [ isSubset pparamUpdateH
+          , isSubset hardForkH
+          , isSubset committeeH
+          , isSubset constitutionH
+          ]
+      )
+      True
+  where
+    pparamUpdateH = hierarchy ^. pfPParamUpdateL . pGraphNodesL
+    hardForkH = hierarchy ^. pfHardForkL . pGraphNodesL
+    committeeH = hierarchy ^. pfCommitteeL . pGraphNodesL
+    constitutionH = hierarchy ^. pfConstitutionL . pGraphNodesL
+    unionGraph =
+      Set.unions
+        [ Set.map unGovPurposeId $ Map.keysSet pparamUpdateH
+        , Set.map unGovPurposeId $ Map.keysSet hardForkH
+        , Set.map unGovPurposeId $ Map.keysSet committeeH
+        , Set.map unGovPurposeId $ Map.keysSet constitutionH
+        ]
+    pparamUpdateRC = roots ^. pfPParamUpdateL . prChildrenL
+    hardForkRC = roots ^. pfHardForkL . prChildrenL
+    committeeRC = roots ^. pfCommitteeL . prChildrenL
+    constitutionRC = roots ^. pfConstitutionL . prChildrenL
+    isSubset h = and $ flip Set.isSubsetOf (Map.keysSet h) . pnChildren <$> Map.elems h
+
+-- | Reconstruct the @`Proposals`@ forest from an @`OMap`@ of
+-- @`GovActionState`@s and the 4 roots (@`PrevGovActionIds`@)
+mkProposals ::
+  (EraPParams era, MonadFail m) =>
+  PrevGovActionIds era ->
+  OMap.OMap (GovActionId (EraCrypto era)) (GovActionState era) ->
+  m (Proposals era)
+mkProposals pgais omap = do
+  ps@(Proposals omap' _roots _hierarchy) <-
+    foldl'
+      ( \mprops gas ->
+          maybe (fail "mkProposals: Could not add a proposal") pure
+            . proposalsAddAction gas
+            =<< mprops
+      )
+      initialProposals
+      omap
+  unless (omap == omap') $ fail "mkProposals: OMap is malformed"
+  unless (proposalsAreConsistent ps) $ fail "mkProposals: Proposals structure is inconsistent"
+  pure ps
+  where
+    initialProposals = pure $ def & pRootsL .~ fromPrevGovActionIds pgais
```

Comment:
Current implementation is too lazy.
With this change the case statement forces the Maybe computation and `foldlM'` will ensure thunks are not built up in on the monadic computation. Also it would be nice to add a little bit more error reporting on problematic GovActionState if one ever occurs.
```suggestion
    foldlM'
      ( \props gas ->
        case proposalsAddAction gas props of
          Nothing -> fail $ "mkProposals: Could not add a proposal " ++ show (gas .^ gasId)
          Just props' -> pure props'
      )
      initialProposals
      omap
  unless (omap == omap') $ fail "mkProposals: OMap is malformed"
  unless (proposalsAreConsistent ps) $ fail "mkProposals: Proposals structure is inconsistent"
  pure ps
  where
    initialProposals = def & pRootsL .~ fromPrevGovActionIds pgais
```